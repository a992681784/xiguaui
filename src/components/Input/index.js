import React, {Component} from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import Icon from '../Icon';

function toThousands(value) {
    return value.replace(/(\d)(?=(?:\d{3})+$)/g, '$1,');
}

const propTypes = {
	className: PropTypes.string,
    labelStyle: PropTypes.object,
    style: PropTypes.object,
    inputStyle: PropTypes.object,
    componentClass: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
        PropTypes.node
    ]),
    pattern: PropTypes.oneOf([
        'inline',
        'vertical',
        'textarea'
    ]),
    border: PropTypes.oneOf([
        'double',
        'bottom',
        'bordered'
    ]),
    onChange: PropTypes.func,
    clear: PropTypes.bool,

};

const defaultProps = {
	labelStyle: {},
	style: {},
	inputStyle: {},
    type: 'text',
    componentClass: 'input',
    pattern: 'inline', //inline,textarea,vertical
    border: 'double', //double bottom bordered
    editable: true,
    clear: true
};

class Input extends Component {
    constructor(props, context){
       super(props, context);
       this.state = {
           value: ''
       }
       this.inputRef = {};

    }

    componentWillMount() {
        let { defaultValue } = this.props;
        if(defaultValue){
            this.setState({
                value: this.formatValue(defaultValue)
            })
        }
    }

    formatValue = (value) => {
        let valueLen = value.length;
        let { type } = this.props;
        switch (type) {
            case 'text':
                break;
            case 'bankCard':
                value = value.replace(/\D/g, '').replace(/(....)(?=.)/g, '$1 ');
                break;
            case 'phone':
                value = value.replace(/\D/g, '').substring(0, 11);

                if (valueLen > 3 && valueLen < 8) {
                    value = `${value.substr(0, 3)} ${value.substr(3)}`;
                } else if (valueLen >= 8) {
                    value = `${value.substr(0, 3)} ${value.substr(3, 4)} ${value.substr(7)}`;
                }
                break;
            case 'number':
                value = value.replace(/\D/g, '');
                break;
            case 'password':
                break;
            case 'money':
                value = value.replace(/\D/g, '');
                value = toThousands(value);
                break;
            default:
                break;
        }
        return value;
    }

    handleChange = (e) => {
        let { onChange, editable } = this.props;
        let value = e.target.value;

        if(editable){
            value = this.formatValue(value);
            this.setState({
                value
            });
            onChange && onChange(value, e)
        }
    }

    handleClear = () => {
        this.setState({
            value: ''
        });
        this.focus();
    }

    focus = () => {
        this.inputRef.focus();
    }

	render() {
		let {
			style,
			className,
			children,
			labelStyle,
			labelClass,
			InputClass,
			inputStyle,
			type,
            componentClass: Component,
			required,
			pattern,
			border,
            editable,
            defaultValue,
			clear,
			...props
		} = this.props;

		let classes = classNames({
			'lebra-input': true,
            [`lebra-input-${pattern}`]: true,
            [`lebra-input-${border}`]: true
		}, className);

		let inputClassName = classNames({
			'lebra-input-item': true
		}, InputClass);

		let labelClassName = classNames({
			'lebra-input-label': true
		}, labelClass);

		if(pattern === 'textarea'){
            Component = pattern;
            if(!props.hasOwnProperty('maxLength')){
                props.maxLength = '140'
            }
        }


		return (
			<div
				className={classes}
				style={ style ? style : null}>
				{ children
					? (<label className={ labelClassName } style={ labelStyle }>{children}</label>)
					: null
				}
				<Component
					type={type}
					style={ inputStyle }
					className={ inputClassName }
                    value={ this.state.value}
                    onChange={ this.handleChange }
					required={required ? required : null}
					{...props}
                    ref={el => this.inputRef = el}
                />
                {
                    pattern !== 'textarea' && clear && this.state.value !== '' ? (
                        <Icon
                            className="lebra-input-clear"
                            type="backspace"
                            onClick={ this.handleClear }
                        />
                    ) : null
                }
                {
                    pattern === 'textarea' ? (
                        <div className="lebra-input-font-length">
                            {this.state.value.length}/
                            <span className="total-length">140</span>
                        </div>
                    ) : null
                }
			</div>
		);
	}
}

Input.propTypes = propTypes;
Input.defaultProps = defaultProps;

export default Input;
