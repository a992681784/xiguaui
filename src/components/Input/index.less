@import "../../style/theme/default";

.lebra-input {
    position: relative;
    min-height: 120px; /*px*/
    background-color: #fff;
    vertical-align: middle;
    margin: 30px 15px;
    overflow: hidden;
    transition: background-color .2s;
    &-label {
        width: 30%;
        margin-left: 16px;
        overflow: hidden;
        padding: 20px 0;
        text-align: left;
        white-space: nowrap;
        color: @inline-label-color;
        font-size: @normal-size;/*px*/
    }
    &-item {
        display: block;
        width: 100%;
        padding: 0;
        box-sizing: border-box;
        background-color: transparent;
        background-image: none;
        color: @font-color;
        border: none;
        outline: 0;
        box-shadow: none;
        border-radius: 0;
        font-size: @normal-size; /*px*/
        line-height: inherit;
        cursor: text;
        &:disabled{
            color: @default-color;
        }
    }
    &-clear{
        position: relative;
        right: 10px;
        top: 2px;
        color: @default-color;
        cursor: pointer;
    }
    &-inline {
        display: flex;
        align-items: center;
        border-top: 1px solid @border-color;/*no*/
        border-bottom: 1px solid @border-color;/*no*/
        .lebra-input-label {

        }
    }
    &-vertical {
        .lebra-input-label {
            display: block;
        }
        .lebra-input-item {
            height: 120px;/*px*/
            padding: 0 16px;
            border-top: 1px solid @border-color;/*no*/
            border-bottom: 1px solid @border-color;/*no*/
        }
    }
    &-textarea {
        .lebra-input-label {
            display: block;
        }
        .lebra-input-item {
            height: 240px;/*px*/
            padding: 40px 16px;
            border: 1px solid @border-color;/*no*/
            resize: none;
            font-family: @font-family;
        }
    }

    &-font-length {
        position: absolute;
        right: 10px;
        bottom: 10px;
        font-size: 14px;/*px*/
        color: @font-color;
        .total-length {
            color: @placeholder-color;
        }
    }
    input::-webkit-input-placeholder, textarea::-webkit-input-placeholder { /* WebKit browsers */
        color: @placeholder-color;
    }
    input::-moz-placeholder, textarea::-moz-placeholder { /* Mozilla Firefox 19 */
        color: @placeholder-color;
    }
    input:-moz-placeholder, textarea:-moz-placeholder { /* Mozilla Firefox 4 to 18 */
        color: @placeholder-color;
    }
    input:-ms-input-placeholder, textarea:-ms-input-placeholder { /* Internet Explorer 10+ */
        color: @placeholder-color;
    }
}
