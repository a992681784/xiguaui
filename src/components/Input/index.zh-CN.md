---
category: Components
type: Data Entry
title: Input
subtitle: 输入框
---

输入框组件，包含了label和input元素。自定义标题和输入样式，完成输入数据校验，支持自定义校验扩展。


## API

适用平台：WEB

|参数|说明|类型|默认值|
|:--|:---|:--|:---|
|style|给最外层元素设置样式|object|{}|
|classNames|给最外层元素设置class|string|''|
|labelStyle|给label元素设置style|object|{}|
|inputStyle|给input元素设置style|object|{}|
|labelClass|给label元素设置class|string|''|
|inputClass|给input元素设置class|string|''|
|type|给input元素设置type|string。type支持text文本，number数字，bankCard银行卡号，phone电话号码|'text'|
|required|给input元素设置required|boolean|-|
|componentClass|给input元素设置required|input/textarea|input|
|placeholder|输入框默认提示信息|string|''|
|editable|是否可编辑|boolean|true|
|disabled|是否可用|boolean|input|
|value|输入框值|string|''|
|defaultValue|输入框默认值|string|''|
|onChange|给input元素设置required|function(e)|-|
|maxLength|输入框可输入最大长度|number|不限制/如果是textarea默认是140|



- 传入的其他props会通过以下方式传递给input元素。

```
let {className, ...others} = this.props;

return (
    <input { ...others } />
)

```

- 获取input的DOM,组件采用如下写法，使用`this.refs.demo.inputRef`获取DOM。

```
...
<Input
    ref="demo"
    placeholder="请输入汉字">
   姓名
</Input>
```

