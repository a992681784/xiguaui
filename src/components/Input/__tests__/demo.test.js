import React from 'react';
import Input from '../index';
import renderer from 'react-test-renderer';

test('Input should be change', () => {
    const component = renderer.create(
        <Input>姓名</Input>
    );
    let input = component.toJSON();
    console.log(input);
    expect(input).toMatchSnapshot();

});
