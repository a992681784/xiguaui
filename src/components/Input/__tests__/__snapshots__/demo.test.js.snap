// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should be change 1`] = `
<div
  className="u-input"
  style={Object {}}
>
  <label
    className="u-input-label"
    style={Object {}}
  >
    姓名
  </label>
  <input
    className="u-input-item"
    required={null}
    style={Object {}}
    type="text"
  />
</div>
`;
